### PR TITLE
added tests for product controller

### PR DESCRIPTION
**Pull request by JuanSaieh**
Changes:
- Added tests for product controller

Trello link: https://trello.com/c/HmMSoooC

### CODE REVIEW

File: spec/controllers/products_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    before do
+      @user = create(:user)
+      @product = create(:product, user_id: @user.id, status: 'published')
+      create(:product, user_id: @user.id)
+    end
+
+    context 'when a user is signed in' do
+      before do
+        sign_in(@user)
+        get :index
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: You can add this line to the before block

File: spec/controllers/products_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+      end
+
+      it 'shows published products and current_user products' do
+        expect(controller.instance_variable_get(:@products)).to match_array(@user.products.or(Product.published))
+      end
+    end
+
+    context 'when there is no user signed in' do
+      before do
+        get :index
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: You are repeating this line twice, you can add it to the before block after the describe.

File: spec/controllers/products_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    it 'assigns a newly created product' do
+      expect(controller.instance_variable_get(:@product)).to be_a_new(Product)
+    end
+  end
+
+  describe 'POST #create' do
+    before do
+      sign_in(@creator)
+    end
+    context 'when product is valid' do
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: Add a line before the context block

File: spec/controllers/products_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,314 @@
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :controller do
+  before do
+    @creator = create(:user)
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: Use a let variable

File: spec/controllers/products_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+      end
+    end
+  end
+
+  describe 'GET #edit' do
+    context 'when user is signed in' do
+      before do
+        sign_in(@creator)
+      end
+      context 'when the user is the owner of the product' do
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: Add a new line here

File: spec/controllers/products_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+      end
+      context 'when user owns the product' do
+        before do
+          put :update, params: { id: Product.first.id, product: { name: 'laptop' } }
+        end
+
+        it 'updates name to laptop' do
+          expect(controller.instance_variable_get(:@product).name).to eq('laptop')
+        end
+        it 'redirects to products_path' do
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: Add a new line here

File: app/controllers/products_controller.rb [hunk 7]
```diff
@@ -5,8 +5,7 @@ class ProductsController < ApplicationController
 
   def index
     if current_user
-      user_products = current_user.products.draft.or(current_user.products.archived)
-      @products = Product.published.or(user_products)
+      @products = Product.published.or(current_user.products)
```
[the remark this answers is not in the document]
Comment (JuanSaieh) [reply app/controllers/products_controller.rb]: Done!

File: spec/controllers/products_controller_spec.rb (hunk 4 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: Done!

File: spec/controllers/products_controller_spec.rb (hunk 1 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: The user needs to sign in before calling the index or the test won't pass!

File: spec/controllers/products_controller_spec.rb (hunk 2 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: The Context 'when a user is signed in' needs to have a logged in user before calling the index, or else the products_controller index method will return only the published products. If we want to add this line to the before block in the describe we would have to sign in before the index call and then log out in this line.

File: spec/controllers/products_controller_spec.rb (hunk 3 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: Done!

File: spec/controllers/products_controller_spec.rb (hunk 5 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: Done!

File: spec/controllers/products_controller_spec.rb (hunk 6 above)
Comment (JuanSaieh) [reply spec/controllers/products_controller_spec.rb]: Done!

File: spec/controllers/products_controller_spec.rb (hunk 1 above)
Comment (Jucerweb) [reply spec/controllers/products_controller_spec.rb]: I meant: `get :index`

File: app/controllers/products_controller.rb [hunk 8]
```diff
@@ -4,12 +4,9 @@ class ProductsController < ApplicationController
   before_action :is_product_from_current_user, only: [:edit, :update, :destroy]
 
   def index
-    if current_user
-      user_products = current_user.products.draft.or(current_user.products.archived)
-      @products = Product.published.or(user_products)
-    else
-      @products = Product.published
-    end
+    byebug
```
Comment (Jucerweb) [inline app/controllers/products_controller.rb]: Byebug

File: spec/controllers/products_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,316 @@
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :controller do
+  let(:creator) { create(:user) }
+  let(:category) { create(:category) }
+  before do
```
Comment (Jucerweb) [inline spec/controllers/products_controller_spec.rb]: Space before the block and after

**Pull request merged**